src/ucp/proto/proto_select.c

```diff
@@ -689,34 +689,34 @@ void ucp_proto_select_short_disable(ucp_proto_select_short_t *proto_short)
     proto_short->rkey_index             = UCP_NULL_RESOURCE;
 }
 
-void
-ucp_proto_select_short_init(ucp_worker_h worker, ucp_proto_select_t *proto_select,
-                            ucp_worker_cfg_index_t ep_cfg_index,
-                            ucp_worker_cfg_index_t rkey_cfg_index,
-                            ucp_operation_id_t op_id, uint32_t op_attr_mask,
-                            unsigned proto_flags,
-                            ucp_proto_select_short_t *proto_short)
+void ucp_proto_select_short_init(ucp_worker_h worker,
+                                 ucp_proto_select_t *proto_select,
+                                 ucp_worker_cfg_index_t ep_cfg_index,
+                                 ucp_worker_cfg_index_t rkey_cfg_index,
+                                 ucp_operation_id_t op_id, unsigned proto_flags,
+                                 ucp_proto_select_short_t *proto_short)
 {
-    ucp_context_h context    = worker->context;
-    const ucp_proto_t *proto = NULL;
+    static const uint32_t op_attributes[] = {0, UCP_OP_ATTR_FLAG_FAST_CMPL,
+                                             UCP_OP_ATTR_FLAG_MULTI_SEND};
+    ucp_context_h context                 = worker->context;
+    const ucp_proto_t *proto              = NULL;
     const ucp_proto_threshold_elem_t *thresh;
     ucp_proto_select_param_t select_param;
     const ucp_proto_single_priv_t *spriv;
     ucp_memory_info_t mem_info;
     ssize_t max_short_signed;
-    uint32_t op_attr;
+    int i;
 
     ucp_memory_info_set_host(&mem_info);
 
     /*
-     * Find the minimal threshold among all protocols for all possible
-     * combinations of bits in 'op_attr_mask'. For example, we are allowed to
-     * use fast-path short protocol only if the message size fits short protocol
-     * in both regular mode and UCP_OP_ATTR_FLAG_FAST_CMPL mode.
+     * Find the minimal threshold for operation 'op_id' among all protocols
+     * with attribute from 'op_attributes'. Fast-path short protocol
+     * can be used only if the message size fits this minimal threshold.
      */
     ucs_log_indent(1);
-    ucs_for_each_submask(op_attr, op_attr_mask) {
-        ucp_proto_select_param_init(&select_param, op_id, op_attr, 0,
+    for (i = 0; i < ucs_static_array_size(op_attributes); ++i) {
```

Comment:
ucs_carray_for_each

Reply:
Will do.

We have many candidates for refactoring in the code...
*search `for i < ucs_static_array_size(`